---
title: "14: RNA-Seq analysis mini-project"
author: "Xinyu Wen (A17115443)"
format: html
toc: TRUE
---
## Backgroud

The data for hand-on session comes from GEO entry: GSE37704, which is associated with the following publication:

> Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.


## Data Import

```{r}
counts <- read.csv("GSE37704_featurecounts.csv", row.names = 1)
colData <- read.csv("GSE37704_metadata.csv")
```

## Inspect and tidy data

Does the `counts` columns match the `colData` rows?
```{r}
head(counts)
head(colData)
```

```{r}
colData$id
```
```{r}
colnames(counts)
```

counts has one extra column "length". Thus we need to remove it
```{r}
countData <- counts[,-1]
head(countData)
```

Check for matching countData and colData
```{r}
colnames(countData) == colData$id
```

> Q1. How many genes in total?

```{r}
nrow(countData)
```

> Q2. Filter to remove zero count genes (rows where there are zero counts in all columns). How many genes are left?

```{r}
to.keep.inds <- which(rowSums(countData) > 0)
new.countData <- countData[to.keep.inds,]
```

```{r}
nrow(new.countData)
```

## Set up DESeq2

```{r}
# message: F
library(DESeq2)
```
Setup input object for DESeq
```{r}
dds <- DESeqDataSetFromMatrix(countData = new.countData,
                              colData = colData,
                              design = ~condition)
```

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
```

## Volcano Plot

```{r}
library(ggplot2)
```

```{r}
mycol <- rep("grey", nrow(res))
mycol[res$log2FoldChange >= 2] <- "indianred"
mycol[res$log2FoldChange <= -2] <- "indianred"
mycol[res$padj > 0.05] <- "grey"
```


```{r}
ggplot(res) +
  aes(log2FoldChange, -log(padj)) +
  geom_point(col = mycol) +
  geom_vline(xintercept = c(-2, 2), col = "blue", lty = 2) +
  geom_hline(yintercept = 0.05, col = "darkgreen", lty = 2)
```


## Gene Annotation
```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```
```{r}
columns(org.Hs.eg.db)
```


```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                      keys=rownames(res),
                      keytype = "ENSEMBL",
                      column = "SYMBOL")

res$genename <- mapIds(org.Hs.eg.db,
                      keys=rownames(res),
                      keytype = "ENSEMBL",
                      column = "GENENAME")

res$entrez <- mapIds(org.Hs.eg.db,
                      keys=rownames(res),
                      keytype = "ENSEMBL",
                      column = "ENTREZID")

```

```{r}
#message: F
library(gage)
library(gageData)
library(pathview)
```

```{r}
data(kegg.sets.hs)
```

```{r}
head(kegg.sets.hs, 3)
```
```{r}
foldchanges <- res$log2FoldChange
```

```{r}
names(foldchanges) <- res$entrez
```


```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
head(keggres$less)
```

Cell cycle figure
```{r}
pathview(foldchanges, pathway.id = "hsa04110")
```
![](hsa04110.pathview.png)
```{r}
pathview(foldchanges, pathway.id = "hsa04620")
```
![](hsa04620.pathview.png)
```{r}
pathview(foldchanges, pathway.id = "hsa00983")
```
![](hsa00983.pathview.png)

Include more figures >3 (Section 2. Pathway Analysis)



## Gene Ontology analysis

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

```{r}
head(gobpres$less)
```
